### PR TITLE
the equivalent APIs should be enabled at most once

### LINKED ISSUE #988: Some kube types exist in more than one group but should be enabled at most once
A number of core types exist in more than one group, e.g. (not intended to be an exhaustive list)

 - `deployments` is found in both `extensions/v1beta1` and `apps/v1`
 - `replicasets` is found in both `extensions/v1beta1` and `deployments.apps/v1`
 - `ingresses` is found in both `extensions/v1beta1` and `networking.k8s.io/v1beta1` (as of Kube 1.14)

Currently a FederatedTypeConfig (FTC) must have a name of the form `<target plural name>.<target group name>`, which allows for one of the above types to be enabled more than once (e.g. an FTC for `deployments.apps` and an FTC for `deployments.extensions`) by varying the group of the generated federated CRD (i.e. by passing `--federated-group` to `kubefedctl enable`). This would result in multiple sync controller targeting different endpoints but potentially fighting to modify the same resources.

Is there a way to mechanically determine that a resource in a given group is the same as in a different group?  Or will it be necessary to hard-code types that appear in multiple groups and only allow enabling federation of a type in a single specific group? e.g.

 - attempting to enable `deployment` would always enable `deployments.apps`
 - attempts to enable `ingress` would enable `ingresses.extensions` pre 1.14 and `ingresses.networking.k8s.io` for 1.14 and greater

/kind bug

### PR DESCRIPTION
**Pull request by xunpan**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #988

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/kubefedctl/enable/deprecatedapis.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+			Version: "v1beta1",
+		},
+		{
+			Group:   "extensions",
+			Version: "v1beta1",
+		},
+	},
+}
+
+func IsEquivalentAPI(existingAPI, newAPI *fedv1b1.APIResource) bool {
```
Comment (marun) [inline pkg/kubefedctl/enable/deprecatedapis.go]: Please add unit test coverage of this function.

File: pkg/kubefedctl/enable/enable.go [hunk 2]
```diff
@@ -286,6 +286,16 @@ func CreateResources(cmdOut io.Writer, config *rest.Config, resources *typeResou
 		}
 
 		for _, ftc := range ftcs.Items {
+			targetAPI := concreteTypeConfig.Spec.TargetType
+			existingAPI := ftc.Spec.TargetType
+			if IsEquivalentAPI(&existingAPI, &targetAPI) {
+				existingName := fmt.Sprintf("%s.%s/%s", existingAPI.PluralName, existingAPI.Group, existingAPI.Version)
+				name := fmt.Sprintf("%s.%s/%s", targetAPI.PluralName, targetAPI.Group, targetAPI.Version)
+
+				return errors.Errorf("Target API %q of FederatedTypeConfig %q is equivalent to existing target API %q of FederatedTypeConfig %q. You can disable existing resource before enabling the new one",
```
Comment (marun) [inline pkg/kubefedctl/enable/enable.go]: -> `("Failed to enable %q. Federation of this type is already enabled for equivalent type %q by FederatedTypeConfig %q", name, existingName, qualifiedFTCName)`

e.g. 

`Failed to enable "deployment.extensions/v1beta1". Federation of this type is already enabled for equivalent type "deployments.apps/v1" by FederatedTypeConfig "kube-federation-system/deployments.apps"`

File: pkg/kubefedctl/enable/enable.go [hunk 3]
```diff
@@ -286,6 +286,16 @@ func CreateResources(cmdOut io.Writer, config *rest.Config, resources *typeResou
 		}
 
 		for _, ftc := range ftcs.Items {
+			targetAPI := concreteTypeConfig.Spec.TargetType
+			existingAPI := ftc.Spec.TargetType
+			if IsEquivalentAPI(&existingAPI, &targetAPI) {
+				existingName := fmt.Sprintf("%s.%s/%s", existingAPI.PluralName, existingAPI.Group, existingAPI.Version)
```
Comment (marun) [inline pkg/kubefedctl/enable/enable.go]: Please use `qualifiedAPIResourceName `

File: pkg/kubefedctl/enable/deprecatedapis.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+import (
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	fedv1b1 "sigs.k8s.io/kubefed/pkg/apis/core/v1beta1"
+)
+
+// Deprecated APIs removed in 1.16 will be served by current equivalent APIs
+// https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/
+//
+// Only allow one of the equivalent APIs for federation to avoid conflict in
```
Comment (marun) [inline pkg/kubefedctl/enable/deprecatedapis.go]: s/conflict in/the possibility of/

**Comment by marun** [comment]
Hmm, or maybe enabling the most recent API is the responsibility of #855?

**Comment by xunpan** [comment]
@marun
comments addressed.

#855 should be used to trace `enabling the most recent API` as you said.

**Comment by marun** [comment]
Thank you!

/lgtm

**Pull request merged**